interview_problems_adarsh/first_missing_positive.py

```diff
@@ -1,94 +0,0 @@
-class Solution:
-    """
-    Class for the solution to LeetCode Problem 41: First Missing Positive
-    """
-
-    def first_missing_positive(self, nums: list[int]) -> int:
-        """
-        Finds and returns the smallest missing positive integer in nums
-
-        :param nums: Unsorted integer array
-        :return: smallest missing positive integer in nums
-        """
-        # change all elements that are out of the possible range of values to 0
-        length = len(nums)
-        for idx in range(length):
-            if nums[idx] < 1 or nums[idx] > length:
-                nums[idx] = 0
-
-        # append 0 to account for the possibility that all values in nums are the smallest possible positive integers, meaning we need to go above those
-        nums.append(0)
-        length += 1
-
-        # helper variable to make it easier to know when to move pointer
-        already_incremented = False
-        # idx will be used when updated certain elements
-        # pointer will be the starting element to update
-        idx = pointer = 0
-        while pointer < length:
-            # if the current element is -1, it has already been updated, so move pointer if possible else return the answer
-            if nums[pointer] == -1:
-                if pointer + 1 < length:
-                    pointer += 1
-                    continue
-                else:
-                    return self.check_smallest(nums)
-            temp = pointer
-            # if the current element is not 0, we use the element at temp as our next index
-            # otherwise, we don't need to use temp, so move pointer if possible else return the answer
-            if nums[temp] != 0:

```

Comment:
I'm having a lot of trouble understanding the algorithm. Need an english-language description of the algorithm in the docstring

 What are `temp` and `pointer` supposed to represent?

We also shouldn't be using `-1` as a dummy value throughout. Is there a better way to do this, with another data structure?